collect_app/src/main/java/org/odk/collect/android/upload/AutoSendWorker.java

```diff
@@ -0,0 +1,322 @@
+     * false otherwise.
+     */
+    private boolean networkTypeMatchesAutoSendSetting(NetworkInfo currentNetworkInfo) {
+        if (currentNetworkInfo == null) {
+            return false;
+        }
+
+        String autosend = (String) GeneralSharedPreferences.getInstance().get(PreferenceKeys.KEY_AUTOSEND);
+        boolean sendwifi = autosend.equals("wifi_only");
+        boolean sendnetwork = autosend.equals("cellular_only");
+        if (autosend.equals("wifi_and_cellular")) {
+            sendwifi = true;
+            sendnetwork = true;
+        }
+
+        return currentNetworkInfo.getType() == ConnectivityManager.TYPE_WIFI
+                && sendwifi || currentNetworkInfo.getType() == ConnectivityManager.TYPE_MOBILE
+                && sendnetwork;
+    }
+
+    /**
+     * Returns instances that need to be auto-sent.
+     */
+    @NonNull
+    private List<Instance> getInstancesToAutoSend(boolean isAutoSendAppSettingEnabled) {
+        InstancesDao dao = new InstancesDao();
+        Cursor c = dao.getFinalizedInstancesCursor();
+        List<Instance> allFinalized = dao.getInstancesFromCursor(c);
+
+        List<Instance> toUpload = new ArrayList<>();
+        for (Instance instance : allFinalized) {
+            if (formShouldBeAutoSent(instance.getJrFormId(), isAutoSendAppSettingEnabled)) {
+                toUpload.add(instance);
+            }
+        }
+
+        return toUpload;
+    }
+
+    /**
+     * Returns whether a form with the specified form_id should be auto-sent given the current
+     * app-level auto-send settings. Returns false if there is no form with the specified form_id.
+     *
+     * A form should be auto-sent if auto-send is on at the app level AND this form doesn't override
+     * auto-send settings OR if auto-send is on at the form-level.
+     *
+     * @param isAutoSendAppSettingEnabled whether the auto-send option is enabled at the app level
+     */
+    public static boolean formShouldBeAutoSent(String jrFormId, boolean isAutoSendAppSettingEnabled) {
+        Cursor cursor = new FormsDao().getFormsCursorForFormId(jrFormId);
+        String formLevelAutoSend = null;
+        if (cursor != null && cursor.moveToFirst()) {
+            try {
+                int autoSendColumnIndex = cursor.getColumnIndex(AUTO_SEND);
+                formLevelAutoSend = cursor.getString(autoSendColumnIndex);
+            } finally {
+                cursor.close();
+            }
+        }
+
+        return formLevelAutoSend == null ? isAutoSendAppSettingEnabled
+                : Boolean.valueOf(formLevelAutoSend);
+    }
+
+    /**
+     * Returns true if at least one form currently on the device specifies that all of its filled
+     * forms should auto-send no matter the connection type.
+     *
+     * TODO: figure out where this should live
+     */
+    private boolean atLeastOneFormSpecifiesAutoSend() {
+        FormsDao dao = new FormsDao();
+        Cursor cursor = dao.getFormsCursor();
+
+        try {
```

Comment:
`try finally` statement can be converted into a `try with resources` statement

Reply:
@grzesiek2010 did mention that elsewhere and although I would like to make that change, I can't get it to compile. https://developer.android.com/studio/write/java8-support#supported_features suggests that Android Studio 3.0 extends support to lower minSdks.

Reply:
Yes, I was able to make the change using Android Studio's auto fix